deps/first/File/src/File.cpp

```diff
@@ -22,8 +24,58 @@ namespace RC::File
         return construct_handle(file_path_and_name, open_properties);
     }
 
+    // BOM sequences
+    const std::string BOM_UTF8 = "\xEF\xBB\xBF";
+    const std::wstring BOM_UTF16_LE = L"\xFEFF";
+    const std::wstring BOM_UTF16_BE = L"\xFFFE";
+
+    // Function to check for BOM and reopen the file as a wide stream
+    auto open_file_skip_BOM(const std::wstring& filename) -> StreamType
+    {
+        std::ifstream file(filename, std::ios::binary);
+        if (!file.is_open())
+        {
+            throw std::runtime_error("Failed to open the file.");
+        }
+
+        char bom[3] = {0};
+        file.read(bom, 3);
+        if (std::string(bom, 3) == BOM_UTF8)
+        {
+            // UTF-8 BOM detected
+            file.close();
+            StreamType wfile(filename);
+            wfile.seekg(3, std::ios::beg); // Skip BOM
+            return wfile;
+        }
+        else
+        {
+            // Check for wide BOM
+            wchar_t wbom;
+            file.seekg(0, std::ios::beg); // Rewind to start
+            file.read(reinterpret_cast<char*>(&wbom), sizeof(wchar_t));
+            file.close();
+            if (wbom == BOM_UTF16_LE[0] || wbom == BOM_UTF16_BE[0])
+            {
+                StreamType wfile(filename);
+#pragma warning(disable : 4996)
+                std::locale loc(wfile.getloc(), new std::codecvt_utf8_utf16<wchar_t>);
```

Comment:
> `std::locale` takes ownership

yes, it's the case. but I don't think we need convert here, as I mentioned before, it's the utf8->utf16 needs a codecvt.. 
because wstring on windows are meant to handle wchar_t, while utf8 uses char.

Reply:
you can refer here https://en.cppreference.com/w/cpp/locale/locale/locale

> Overload (7) is typically called with its second argument, f, obtained directly from a new-expression: the locale is responsible for calling the matching delete from its own destructor.

Reply:
This is good information, thanks.
It's odd to me that it needs to be heap allocated, but if that's the requirement then that's what we'll do.
@narknon Ignore my suggestion about passing the address above, that won't work.

Reply:
This would be worse because the locale would try to manage the lifetime of a stack allocated object

Reply:
As I said, ignore my suggestion.

Reply:
Sorry, hadn't refreshed the page so I didn't know Yang responded